# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/clients/manager/ClientManager.cpp
# Commit: 6125bb02952a
# Full Hash: 6125bb02952aff71c0d757e238a7752101b2d6c2
# Author: Jean-Yves Avenard <jyavenard@mozilla.com>
# Date: 2019-09-03 04:06:20
# Regressor Bug: 1556489
# File Overlap Count: 1
# Description:
#   Bug 1556489 - P23 - Only create new ClientSource objects in the content, but use the recreated info from redirects in the parent. r=asuth
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D42260
# ==============================================================================

diff -r 06529d12b589 -r 6125bb02952a dom/clients/manager/ClientManager.cpp
--- a/dom/clients/manager/ClientManager.cpp	Mon Aug 26 12:18:57 2019 +1000
+++ b/dom/clients/manager/ClientManager.cpp	Tue Aug 27 08:53:37 2019 +1000
@@ -132,6 +132,25 @@
   return source;
 }
 
+UniquePtr<ClientSource> ClientManager::CreateSourceInternal(
+    const ClientInfo& aClientInfo, nsISerialEventTarget* aEventTarget) {
+  NS_ASSERT_OWNINGTHREAD(ClientManager);
+
+  ClientSourceConstructorArgs args(aClientInfo.Id(), aClientInfo.Type(),
+                                   aClientInfo.PrincipalInfo(),
+                                   aClientInfo.CreationTime());
+  UniquePtr<ClientSource> source(new ClientSource(this, aEventTarget, args));
+
+  if (IsShutdown()) {
+    source->Shutdown();
+    return source;
+  }
+
+  source->Activate(GetActor());
+
+  return source;
+}
+
 already_AddRefed<ClientHandle> ClientManager::CreateHandleInternal(
     const ClientInfo& aClientInfo, nsISerialEventTarget* aSerialEventTarget) {
   NS_ASSERT_OWNINGTHREAD(ClientManager);
@@ -255,6 +274,34 @@
 }
 
 // static
+UniquePtr<ClientSource> ClientManager::CreateSourceFromInfo(
+    const ClientInfo& aClientInfo, nsISerialEventTarget* aEventTarget) {
+  RefPtr<ClientManager> mgr = GetOrCreateForCurrentThread();
+  return mgr->CreateSourceInternal(aClientInfo, aEventTarget);
+}
+
+Maybe<ClientInfo> ClientManager::CreateInfo(ClientType aType,
+                                            nsIPrincipal* aPrincipal) {
+  MOZ_ASSERT(NS_IsMainThread());
+  MOZ_ASSERT(aPrincipal);
+
+  PrincipalInfo principalInfo;
+  nsresult rv = PrincipalToPrincipalInfo(aPrincipal, &principalInfo);
+  if (NS_WARN_IF(NS_FAILED(rv))) {
+    MOZ_CRASH("ClientManager::CreateSource() cannot serialize bad principal");
+  }
+
+  nsID id;
+  rv = nsContentUtils::GenerateUUIDInPlace(id);
+  MOZ_DIAGNOSTIC_ASSERT(NS_SUCCEEDED(rv));
+  if (NS_WARN_IF(NS_FAILED(rv))) {
+    return Nothing();
+  }
+
+  return Some(ClientInfo(id, aType, principalInfo, TimeStamp::Now()));
+}
+
+// static
 already_AddRefed<ClientHandle> ClientManager::CreateHandle(
     const ClientInfo& aClientInfo, nsISerialEventTarget* aSerialEventTarget) {
   RefPtr<ClientManager> mgr = GetOrCreateForCurrentThread();